### PR TITLE
Bug 1875817 - Translations Integration Toolbar Showing

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1875817

### CODE REVIEW

File: fenix/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -242,6 +242,10 @@
     <string name="browser_toolbar_erase">Erase browsing history</string>
     <!-- Content description for the translate page toolbar button that opens the translations dialog when no translation has occurred. -->
     <string name="browser_toolbar_translate">Translate page</string>
+    <!-- Content description (not visible, for screen readers etc.) for the translate page toolbar button that opens the translations dialog when the page is translated successfully.
+          The first parameter is the name of the language from which the page was translated.(for example: English)
```
Comment (ohall-m) [inline fenix/app/src/main/res/values/strings.xml:246]: nit:
```suggestion
          The first parameter is the name of the language from which the page was translated.(For example: English)
```

File: fenix/app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -242,6 +242,10 @@
...
+          The second parameter is the name of the language into which the page was translated.(for example: France) -->
```
Comment (ohall-m) [inline fenix/app/src/main/res/values/strings.xml:247]: nit: 
```suggestion
          The second parameter is the name of the language into which the page was translated.(For example: France) -->
```

File: fenix/app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -242,6 +242,10 @@
...
+    <string name="browser_toolbar_translate_successfully">Page translated from %1$s to %2$s.</string>
```
Comment (ohall-m) [inline fenix/app/src/main/res/values/strings.xml:248]: nit:
```suggestion
    <string name="browser_toolbar_translated_successfully">Page translated from %1$s to %2$s.</string>
```

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt [hunk 4]
```diff
@@ -236,32 +238,56 @@ class BrowserFragment : BaseBrowserFragment(), UserInteractionHandler {
...
-                },
-                listener = { browserToolbarInteractor.onTranslationsButtonClicked() },
-            )
-
+        val translationsAction = Toolbar.ActionButton(
+            AppCompatResources.getDrawable(
+                context,
+                R.drawable.mozac_ic_translate_24,
+            ),
+            contentDescription = context.getString(R.string.browser_toolbar_translate),
```
Comment (ohall-m) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:251]: This sounded right in TalkBack and the color change both tested well for me when checking states!

File: fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            ).joinBlocking()
+
+            browserStore.dispatch(
+                TranslationsAction.TranslateSuccessAction(
+                    tabId = tab.id,
+                    operation = TranslationOperation.TRANSLATE,
+                ),
+            ).joinBlocking()
+
+            Mockito.verify(onIconChanged).invoke(
```
Comment (ohall-m) [inline fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt:105]: Thanks for the tests! Could you also add the parameter names here and the tests below to improve readability?

**Review by ohall-m — APPROVED** [review]
Thanks for working on this. It is a really great refinement!

I also tested against having an auto-translated language set too and it worked well in that case too. Worked great overall!

* It looks like this bug fulfills [1874477](https://bugzilla.mozilla.org/show_bug.cgi?id=1874477) too. I think that is good, but we need to close or update 1874477 based on how much is achieved here.
* I mentioned a few visual refinements around the icon. If they are more than a quick adjustment, we can put them on 1874477 or whatever you prefer.

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+                if (fromSelected != null && toSelected != null) {
+                    onIconChanged(
+                        true,
+                        true,
+                        fromSelected,
+                        toSelected,
+                    )
+                }
+            } else if (translationsState.isExpectedTranslate) {
```
Comment (ohall-m) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:59]: (I spotted an issue on the AC side for `isExpectedTranslate`, it doesn't impact the way this patch works, but I'm going to prioritize bug [1881294](https://bugzilla.mozilla.org/show_bug.cgi?id=1881294) on the AC side to clear the state.)

File: fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            )
+            binding.start()
+
+            browserStore.dispatch(
+                TranslationsAction.TranslateExpectedAction(
+                    tabId = tabId,
+                ),
+            ).joinBlocking()
+
+            Mockito.verify(onIconChanged).invoke(
```
Comment (gabrielluong) [inline fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt:136]:
```suggestion
           verify(onIconChanged).invoke(
```
Could you import `Mockito.verify` so we can drop `Mockito` here

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                val toSelected = translationsState.translationEngineState?.initialToLanguage(
+                    translationsState.supportedLanguages?.toLanguages,
+                )
+
+                if (fromSelected != null && toSelected != null) {
+                    onIconChanged(
+                        true,
+                        true,
+                        fromSelected,
+                        toSelected,
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:56]: Let's add the named param here to make it easier for future refactoring

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            } else if (translationsState.isExpectedTranslate) {
+                onIconChanged(
+                    true,
+                    false,
+                    null,
+                    null,
+                )
+            } else {
+                onIconChanged(false, false, null, null)
+            }
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:68]: Let's add the named params

File: android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt [hunk 10]
```diff
@@ -279,26 +279,46 @@ interface Toolbar : ScrollableToolbar {
...
 
-            imageButton.setBackgroundResource(backgroundResource)
-            padding?.let { imageButton.setPadding(it) }
+        /**
+         * Changes the content description and the tint colour of the view.
+         *
+         * @param contentDescription The content description to use.
+         * @param tintColorResource ID of color resource to tint the icon.
+         */
+        fun updateView(
```
Comment (gabrielluong) [inline android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt:314]: Please add tests for this and a changelog entry. We're essentially exposing a function that allows for callback or fine-grain control of updating the view. Normally, we would expect a toggle button to immediate just change colors if it was selected or not. However, I am guessing the reasoning you went with exposing this fine-grain control is to ensure that the translation processing went through and to display an accurate status. Could you confirm this was the intention?

File: android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt [hunk 11]
```diff
@@ -279,26 +279,46 @@ interface Toolbar : ScrollableToolbar {
...
     ) : Action {
+        private var view: WeakReference<AppCompatImageButton>? = null
+
+        override fun createView(parent: ViewGroup): View =
+            AppCompatImageButton(parent.context).also { imageButton ->
+                view = WeakReference(imageButton)
+
+                imageButton.setImageDrawable(imageDrawable)
+                imageButton.contentDescription = contentDescription
+                imageButton.setTintResource(iconTintColorResource)
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt:290]: Done

File: fenix/app/src/main/res/values/strings.xml (hunk 1 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/res/values/strings.xml:246]: Done

File: fenix/app/src/main/res/values/strings.xml (hunk 3 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/res/values/strings.xml:248]: Done

File: fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt (hunk 7 above)
Comment (iorgamgabriel) [reply fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt:136]: Done

File: fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt (hunk 5 above)
Comment (iorgamgabriel) [reply fenix/app/src/test/java/org/mozilla/fenix/browser/TranslationsIconBindingTest.kt:105]: I can't add names to the callback invoke function.

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt (hunk 8 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:56]: I can't add names to the callback invoke function.

File: fenix/app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -242,6 +242,10 @@
     <string name="browser_toolbar_erase">Erase browsing history</string>
     <!-- Content description for the translate page toolbar button that opens the translations dialog when no translation has occurred. -->
     <string name="browser_toolbar_translate">Translate page</string>
+    <!-- Content description (not visible, for screen readers etc.) for the translate page toolbar button that opens the translations dialog when the page is translated successfully.
+          The first parameter is the name of the language from which the page was translated.(For example: English)
+          The second parameter is the name of the language into which the page was translated.(For example: France) -->
```
Comment (gabrielluong) [inline fenix/app/src/main/res/values/strings.xml:247]:
```suggestion
          The first parameter is the name of the language that is displayed in the original page. (For example: English)
          The second parameter is the name of the language which the page was translated to. (For example: French) -->
```

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt [hunk 13]
```diff
@@ -569,9 +597,11 @@ class BrowserFragment : BaseBrowserFragment(), UserInteractionHandler {
                     isNewCollection -> {
                         R.string.create_collection_tabs_saved_new_collection
                     }
+
                     tabSize > 1 -> {
                         R.string.create_collection_tabs_saved
                     }
+
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:604]: Let's undo this and avoid adding unrelated changes

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import mozilla.components.concept.engine.translate.initialFromLanguage
+import mozilla.components.concept.engine.translate.initialToLanguage
+import mozilla.components.lib.state.helpers.AbstractBinding
+
+/**
+ * Helper for observing Translation state from browserStore.
+ * @param browserStore Containing the information about [TranslationsState].
+ * @param sessionId Current open tab session id.
+ * @param onIconChanged Invoked when icon visibility, "from" language selected,
+ * "to" language selected, and the translated flag are changing.
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:24]:
```suggestion
 * A binding for observing [TranslationsState] changes from the [BrowserStore] and updating the translations action button.
 *
 * @param browserStore [BrowserStore] observed for any changes related to [TranslationsState].
 * @param sessionId Current open tab session id.
 * @param onIconChanged Invoked when the translations action button should be updated with the new translations state. 
```

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+ */
+class TranslationsIconBinding(
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:26]:
```suggestion
class TranslationsActionBinding(
```

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt (hunk 8 above)
Comment (gabrielluong) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:56]: Okay.

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    private val browserStore: BrowserStore,
+    private val sessionId: String,
+    private val onIconChanged: (
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:29]:
```suggestion
    private val onStateUpdated: (
```
Just thinking about the naming of this a bit more - while the icon will be changed depending on how the consumer uses this lambda, this binding class is actually pretty generic and only serving as a callback passing in the new translation state information to the consumer. So, I think we should actually remove any mention of `icon` and treat this as reusable for anything that cares about reacting to these state changes.

File: android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt (hunk 10 above)
Comment (gabrielluong) [reply android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt:314]: I am not quite sure I follow. In [updateView](https://github.com/mozilla-mobile/firefox-android/pull/5600/files#diff-d6281fdb81a5995139750c3bb8316ab4494ebfac844287ff0c691bf805661e81R269), we are changing the tint colours based on the translation state. The toggle buttons would allow you to similarly sent the tint for the selected or unselected state.

Comment (iorgamgabriel) [reply android-components/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt:314]: Yes, but the colour is changing when the user presses the icon. 
    /**
         * Changes the selected state of the action to the inverse of its current state.
         *
         * @param notifyListener If true (default) the listener will be notified about the state change.
         */
        fun toggle(notifyListener: Boolean = true) {
            setSelected(!selected, notifyListener)
        }

File: fenix/app/src/main/res/values/strings.xml (hunk 12 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/res/values/strings.xml:247]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt (hunk 13 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:604]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt (hunk 14 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:24]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt (hunk 15 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:26]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt (hunk 16 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt:29]: Done

**Review by gabrielluong — APPROVED** [review]
